File: bigseqkit/head_genome.go
package bigseqkit

import (
	"github.com/shenwei356/bio/seqio/fastx"
	"github.com/shenwei356/util/stringutil"
	"ignis/driver/api"
	"strings"
)

type SeqKitHeadGenomeOptions struct {
	inner HeadGenomeOptions
}

type HeadGenomeOptions struct {
	Config          KitConfig
	MiniCommonWords *int64
}

func (this *HeadGenomeOptions) setDefaults() *HeadGenomeOptions {
	this.Config.setDefaults()
	setDefault(&this.MiniCommonWords, 1)

	return this
}

func (this *SeqKitHeadGenomeOptions) Config(v *SeqKitConfig) *SeqKitHeadGenomeOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitHeadGenomeOptions) MiniCommonWords(v int64) *SeqKitHeadGenomeOptions {
	this.inner.MiniCommonWords = &v
	return this
}

func HeadGenome(input *api.IDataFrame[string], o *SeqKitHeadGenomeOptions) (*api.IDataFrame[string], error) {
	if o == nil {
		o = &SeqKitHeadGenomeOptions{}
	}
	opts := o.inner
	opts.setDefaults()

	firstSeq, err := input.Take(1)
	if err != nil {
		return nil, err
	}
	fastxReader, err := fastx.NewReaderFromIO(nil, strings.NewReader(firstSeq[0]), *opts.Config.IDRegexp)
	if err != nil {
		return nil, err
	}
	record, err := fastxReader.Read()
	if err != nil {
		return nil, err
	}

	prefixes := stringutil.Split(string(record.Desc), "\t ")

	lib, err := api.AddParam(libSource("HeadGenome"), "opts", OptionsToString(opts))
	if err != nil {
		return nil, err
	}
	lib, err = api.AddParam(lib, "prefixes", prefixes)
	if err != nil {
		return nil, err
	}

	return api.MapPartitionsWithIndex[string, string](input, lib)
}

File: bigseqkit-py/bigseqkit/head_genome.py
from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitHeadGenomeOptions:

    def __init__(self):
        self.__inner = HeadGenomeOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def headGenomeOptions(self, v: int):
        self.__inner.HeadGenomeOptions = v

    def _run(self, input: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()

        firstSeq = input.take(1)

        header = firstSeq.split("\n")[0]
        prefixes = header.split()[1:]

        lib = _libSource("HeadGenome")\
            .addParam("opts", _optionsToString(opts))\
            .addParam("prefixes", prefixes)

        return input.mapPartitionsWithIndex(lib)


class HeadGenomeOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.HeadGenomeOptions = None  # int

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "HeadGenomeOptions", 1)


def headGenome(input: IDataFrame, o: SeqKitHeadGenomeOptions = None, **kwargs):
    if o is None:
        o = SeqKitHeadGenomeOptions()
    return o._run(input, **kwargs)
